source/common/router/router.cc

```diff
@@ -621,6 +594,65 @@ void Filter::onUpstreamReset(UpstreamResetType type,
   }
 }
 
+bool Filter::maybeRetryReset(const absl::optional<Http::StreamResetReason>& reset_reason) {
+  Upstream::HostDescriptionConstSharedPtr upstream_host;
+  if (upstream_request_) {
+    upstream_host = upstream_request_->upstream_host_;
+  }
+
+  // We don't retry on a global timeout or if we already started the response.
+  if (!downstream_response_started_ && retry_state_) {
+    // Notify retry modifiers about the attempted host.
+    if (upstream_host != nullptr) {
+      retry_state_->onHostAttempted(upstream_host);
+    }
+
+    RetryStatus retry_status =
+        retry_state_->shouldRetry(nullptr, reset_reason, [this]() -> void { doRetry(); });
+    if (retry_status == RetryStatus::Yes && setupRetry(true)) {
+      if (upstream_host) {
+        upstream_host->stats().rq_error_.inc();
+      }
+      return true;
+    } else if (retry_status == RetryStatus::NoOverflow) {
+      callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::UpstreamOverflow);
+    } else if (retry_status == RetryStatus::NoRetryLimitExceeded) {
+      callbacks_->streamInfo().setResponseFlag(
+          StreamInfo::ResponseFlag::UpstreamRetryLimitExceeded);
+    }
+  }
+
+  return false;
+}
+
+void Filter::onUpstreamReset(const absl::optional<Http::StreamResetReason>& reset_reason) {
+  ASSERT(upstream_request_);
+  ENVOY_STREAM_LOG(debug, "upstream reset: reset reason {}", *callbacks_,
+                   reset_reason ? Http::Utility::resetReasonToString(reset_reason.value()) : "");
+
+  Upstream::HostDescriptionConstSharedPtr upstream_host;
+  if (upstream_request_) {
+    upstream_host = upstream_request_->upstream_host_;
+  }
+
+  Http::Code code = Http::Code::ServiceUnavailable;
+
+  updateOutlierDetection(code);
+
+  if (maybeRetryReset(reset_reason)) {
+    return;
+  }
+
+  StreamInfo::ResponseFlag response_flags =
+      streamResetReasonToResponseFlag(reset_reason.value());
```

Comment:
This blows up if you don't have a `reset_reason` since you never check it anywhere. If you know that you always have one I would make `onUpstreamReset` just take the reason directly instead of wrapping it in optional